### PR TITLE
WIP: JSDK wire enableAutomaticSubscription

### PR DESCRIPTION
**Pull request by makarandp0**
This is Work in Progress. This changes adds a new property `enableAutomaticSubscription` in ConnectOptions.  When set to `false` ( default is `true`), this will cause local participant to not subscribe to any of the remote tracks. 

~This PR also adds a small (unrelated) change where instead of returning a connect promise with rejection handler attached, it generates a new promise - This will ensure that user will get `unhandled rejection` if they forget to handle the rejection on the returned promise.~ 

TODO:
- [x] add integration tests

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -6,6 +6,21 @@ For 1.x changes, go [here](https://github.com/twilio/twilio-video.js/blob/suppor
 New Features
 ------------
 
+- Added a new property `enableAutomaticSubscription` to control Track subscription behavior in Group Rooms.
```
Comment (manjeshbhargav) [inline CHANGELOG.md]: @makarandp0 How about:

> By default, you will subscribe to all RemoteTracks shared by other Participants in a Room. You can now override this behavior through a new ConnectOptions flag `automaticSubscription`. Setting it to `false` will make sure that you *will not* subscribe to any RemoteTrack in a Group or Small Group Room. Setting it to `true`, or not setting it at all preserves the default behavior. This flag does not have any effect in a Peer-to-Peer Room. (JSDK-2395)

```js
const { connect } = require('twilio-video');
const room = await connect(token, {
  automaticSubscription: false
});
```

?

File: CHANGELOG.md [hunk 2]
```diff
@@ -6,6 +6,21 @@ For 1.x changes, go [here](https://github.com/twilio/twilio-video.js/blob/suppor
...
+  - Selecting `false` has no impact in a Peer-to-Peer Room
+
+```js
+  const { connect } = require('twilio-video');
+  const room = await connect(token, {
+    enableAutomaticSubscription: false
+  });
+```
+
+New Features
```
Comment (manjeshbhargav) [inline CHANGELOG.md]: @makarandp0 Duplicate header.

File: lib/connect.js [hunk 3]
```diff
@@ -125,7 +125,7 @@ function connect(token, options) {
     abortOnIceServersTimeout: false,
     createLocalTracks,
     dominantSpeaker: false,
-    networkQuality: false,
+    enableAutomaticSubscription: true,
```
Comment (manjeshbhargav) [inline lib/connect.js]: @makarandp0 `automaticSubscription`.

File: lib/connect.js [hunk 4]
```diff
@@ -260,9 +261,8 @@ function connect(token, options) {
     } else {
       log.info('Error while connecting to a Room:', error);
     }
+    throw error;
   });
-
-  return cancelableRoomPromise;
```
Comment (manjeshbhargav) [inline lib/connect.js]: @makarandp0 Can you revert this change? We need to return a `CancelablePromise` here. Your change will return a regular `Promise` I think.

File: lib/connect.js [hunk 5]
```diff
@@ -278,6 +278,10 @@ function connect(token, options) {
  * @property {boolean|CreateLocalTrackOptions} [audio=true] - Whether or not to
  *   get local audio with <code>getUserMedia</code> when <code>tracks</code>
  *   are not provided.
+ * @property {boolean} [enableAutomaticSubscription=true] - By default,
```
Comment (manjeshbhargav) [inline lib/connect.js]: @makarandp0 Can you replicate my changelog entry suggestion here? Also, `automaticSubscription`.

File: lib/signaling/v2/cancelableroomsignalingpromise.js [hunk 6]
```diff
@@ -58,6 +58,7 @@ function createCancelableRoomSignalingPromise(token, wsServer, localParticipant,
         transportOptions = Object.assign({
           dominantSpeaker: options.dominantSpeaker,
           environment: options.environment,
+          enableAutomaticSubscription: options.enableAutomaticSubscription,
```
Comment (manjeshbhargav) [inline lib/signaling/v2/cancelableroomsignalingpromise.js]: @makarandp0 `automaticSubscription`.

File: lib/signaling/v2/twilioconnectiontransport.js [hunk 7]
```diff
@@ -118,6 +119,9 @@ class TwilioConnectionTransport extends StateMachine {
           options.realm,
           eventPublisherOptions)
       },
+      _enableAutomaticSubscription: {
+        value: options.enableAutomaticSubscription
```
Comment (manjeshbhargav) [inline lib/signaling/v2/twilioconnectiontransport.js]: @makarandp0 `automaticSubscription`.

File: lib/util/index.js [hunk 8]
```diff
@@ -541,6 +541,21 @@ function createMediaSignalingPayload(dominantSpeaker, networkQuality) {
       : {});
 }
 
+/**
+ * Create the Subscribe payload included in an RSP connect message.
+ * @param {boolean} enableAutomaticSubscription - whether to subscribe to all tracks
+ * @returns {object}
+ */
+function createSubscribePayload(enableAutomaticSubscription) {
```
Comment (manjeshbhargav) [inline lib/util/index.js]: @makarandp0 `automaticSubscription`.

File: lib/connect.js (hunk 4 above)
Comment (makarandp0) [reply lib/connect.js]: Good point about `CancelablePromise`, I will revert this change, We can take/discuss this separately.

**Comment by makarandp0** [comment]
@manjeshbhargav, Thank you for the review, I have incorporated your feedback, and added integration tests. Can you take another look.

File: CHANGELOG.md [hunk 9]
```diff
@@ -6,6 +6,19 @@ For 1.x changes, go [here](https://github.com/twilio/twilio-video.js/blob/suppor
 New Features
 ------------
 
+- By default, you will subscribe to all RemoteTracks shared by other Participants in a Room.
+  You can now override this behavior through a new ConnectOptions flag automaticSubscription.
```
Comment (manjeshbhargav) [inline CHANGELOG.md]: @makarandp0 s/automaticSubscription/\`automaticSubscription\`/g

File: CHANGELOG.md [hunk 10]
```diff
@@ -6,6 +6,19 @@ For 1.x changes, go [here](https://github.com/twilio/twilio-video.js/blob/suppor
...
+  Setting it to false will make sure that you will not subscribe to any RemoteTrack in a Group or
```
Comment (manjeshbhargav) [inline CHANGELOG.md]: @makarandp0 s/false/\`false\`/g

File: CHANGELOG.md [hunk 11]
```diff
@@ -6,6 +6,19 @@ For 1.x changes, go [here](https://github.com/twilio/twilio-video.js/blob/suppor
...
+  Small Group Room. Setting it to true, or not setting it at all preserves the default behavior.
```
Comment (manjeshbhargav) [inline CHANGELOG.md]: @makarandp0 s/false/\`true\`/g

File: lib/connect.js [hunk 12]
```diff
@@ -125,7 +125,7 @@ function connect(token, options) {
     abortOnIceServersTimeout: false,
     createLocalTracks,
     dominantSpeaker: false,
-    networkQuality: false,
+    automaticSubscription: true,
```
Comment (manjeshbhargav) [inline lib/connect.js]: @makarandp0 Nitpick: Can you move this flag just above `createLocalTracks`?

File: lib/connect.js [hunk 13]
```diff
@@ -278,6 +279,12 @@ function connect(token, options) {
  * @property {boolean|CreateLocalTrackOptions} [audio=true] - Whether or not to
  *   get local audio with <code>getUserMedia</code> when <code>tracks</code>
  *   are not provided.
+ * @property {boolean} [automaticSubscription=true] - By default, you will subscribe
+ *   to all RemoteTracks shared by other Participants in a Room. You can now override this
```
Comment (manjeshbhargav) [inline lib/connect.js]: @makarandp0 

> You can now override this behavior by setting this flag to false. This will make sure that you will not subscribe to any RemoteTrack in a Group or Small Group Room. ...

File: lib/signaling/v2/twilioconnectiontransport.js [hunk 14]
```diff
@@ -118,6 +119,9 @@ class TwilioConnectionTransport extends StateMachine {
           options.realm,
           eventPublisherOptions)
       },
+      _automaticSubscription: {
```
Comment (manjeshbhargav) [inline lib/signaling/v2/twilioconnectiontransport.js]: @makarandp0 Nitpick: Can you move this above `_dominantSpeaker` and below `_accessToken`?

File: lib/util/index.js [hunk 15]
```diff
@@ -541,6 +541,21 @@ function createMediaSignalingPayload(dominantSpeaker, networkQuality) {
       : {});
 }
 
+/**
+ * Create the Subscribe payload included in an RSP connect message.
```
Comment (manjeshbhargav) [inline lib/util/index.js]: @makarandp0 In the future, this can also be part of an RSP "update" message. So, can you say "... in an RSP connect/update message?

File: test/integration/spec/connect.js [hunk 16]
```diff
@@ -104,6 +104,49 @@ describe('connect', function() {
     });
   });
 
+  describe.only('automaticSubscription option', () => {
```
Comment (manjeshbhargav) [inline test/integration/spec/connect.js]: @makarandp0 Remove `.only`.

File: test/integration/spec/connect.js [hunk 17]
```diff
@@ -104,6 +104,49 @@ describe('connect', function() {
...
+    ['missing', 'enabled', 'disabled'].forEach((option) => {
```
Comment (manjeshbhargav) [inline test/integration/spec/connect.js]: @makarandp0 Nitpick: Remove `()` around `option`.

File: test/integration/spec/connect.js [hunk 18]
```diff
@@ -104,6 +104,49 @@ describe('connect', function() {
...
+        let notsubscribed = 0;
+        flatMap(thoseParticipants, participant => [...participant.tracks.values()]).forEach(publication => {
+          if (publication.isSubscribed) {
+            subscribed++;
+          } else {
+            notsubscribed++;
+          }
+        });
+        if (expectSubscriptionsEnabled) {
+          assert.equal(subscribed, 4, 'subscribed should be 4, but were: ' + subscribed);
```
Comment (manjeshbhargav) [inline test/integration/spec/connect.js]: @makarandp0 Nitpick: Use \`\` string for error message here.

File: test/integration/spec/connect.js [hunk 19]
```diff
@@ -104,6 +104,49 @@ describe('connect', function() {
...
+          assert.equal(notsubscribed, 0, 'notsubscribed should be zero, but were: ' + notsubscribed);
```
Comment (manjeshbhargav) [inline test/integration/spec/connect.js]: @makarandp0 Nitpick: Use \`\` string for error message here.

File: test/integration/spec/connect.js [hunk 20]
```diff
@@ -104,6 +104,49 @@ describe('connect', function() {
...
+        } else {
+          assert.equal(subscribed, 0, 'subscribed should be 4, but were: ' + subscribed);
```
Comment (manjeshbhargav) [inline test/integration/spec/connect.js]: @makarandp0 Nitpick: Use \`\` string for error message here.

File: test/integration/spec/connect.js [hunk 21]
```diff
@@ -104,6 +104,49 @@ describe('connect', function() {
...
+          assert.equal(notsubscribed, 4, 'notsubscribed should be zero, but were: ' + notsubscribed);
```
Comment (manjeshbhargav) [inline test/integration/spec/connect.js]: @makarandp0 Nitpick: Use \`\` string for error message here.

File: test/integration/spec/connect.js (hunk 16 above)
Comment (manjeshbhargav) [reply test/integration/spec/connect.js]: @makarandp0 Also, just "automaticSubscription".

**Comment by makarandp0** [comment]
This change is now part of https://github.com/twilio/twilio-video.js/pull/668
